include/kernels/dslash_ndeg_twisted_mass_preconditioned.cuh

```diff
@@ -0,0 +1,144 @@
+
+    if (xpay && kernel_type == INTERIOR_KERNEL) {
+
+      if (!dagger || asymmetric) { // apply inverse twist which is undone below
+        // use consistent load order across s to ensure better cache locality
+        Vector x0 = arg.x(x_cb + 0*arg.dc.volume_4d_cb, my_spinor_parity);
+        Vector x1 = arg.x(x_cb + 1*arg.dc.volume_4d_cb, my_spinor_parity);
+        if (flavor==0) out += arg.a_inv * (x0 + arg.b_inv*x0.igamma(4) + arg.c_inv*x1);
+        else           out += arg.a_inv * (x1 - arg.b_inv*x1.igamma(4) + arg.c_inv*x0);
+      } else {
+        Vector x = arg.x(my_flavor_idx, my_spinor_parity);
+        out += x; // just directly add since twist already applied in the dslash
+      }
+
+    } else if (kernel_type != INTERIOR_KERNEL && active) {
+      // if we're not the interior kernel, then we must sum the partial
+      Vector x = arg.out(my_flavor_idx, my_spinor_parity);
+      out += x;
+    }
+
+    if ( isComplete<kernel_type>(arg, coord) && active ) {
+      if (!dagger || asymmetric) { // apply A^{-1} to D*in
+        VectorCache<real,Vector> cache;
+        // to apply the preconditioner we need to put "out" in shared memory so the other flavor can access it
+        cache.save(out);
+        cache.sync(); // safe to sync in here since other threads will exit
+
+        if (flavor==0) out = arg.a * (out + arg.b*out.igamma(4) + arg.c*cache.load(threadIdx.x,1,threadIdx.z));
+        else           out = arg.a * (out - arg.b*out.igamma(4) + arg.c*cache.load(threadIdx.x,0,threadIdx.z));
+      }
+    }
+
+    if (kernel_type != EXTERIOR_KERNEL_ALL || active) arg.out(my_flavor_idx, my_spinor_parity) = out;
+  }
+
+  // CPU kernel for applying the non-degenerate twisted-mass operator to a vector
+  template <typename Float, int nDim, int nColor, int nParity, bool dagger, bool asymmetric, bool xpay, KernelType kernel_type, typename Arg>
+  void ndegTwistedMassCPU(Arg arg)
+  {
+    for (int parity= 0; parity < nParity; parity++) {
+      // for full fields then set parity from loop else use arg setting
+      parity = nParity == 2 ? parity : arg.parity;
+
+      for (int x_cb = 0; x_cb < arg.threads; x_cb++) { // 4-d volume
+        for (int flavor=0; flavor<2; flavor++) {
+          ndegTwistedMass<Float,nDim,nColor,nParity,dagger,asymmetric,xpay,kernel_type>(arg, x_cb, flavor, parity);
+        }
+      } // 4-d volumeCB
+    } // parity
+
+  }
+
+  // GPU Kernel for applying the non-dengenerate twisted-mass operator to a vector
+  template <typename Float, int nDim, int nColor, int nParity, bool dagger, bool asymmetric, bool xpay, KernelType kernel_type, typename Arg>
+  __global__ void ndegTwistedMassGPU(Arg arg)
+  {
+    int x_cb = blockIdx.x*blockDim.x + threadIdx.x;
+    if (x_cb >= arg.threads) return;
+
+    // for this operator flavor can be spread onto different blocks
+    int flavor = blockIdx.y*blockDim.y + threadIdx.y;
+
+    // for full fields set parity from y thread index else use arg setting
```

Comment:
typo --- `z` thread index. this probably got copy+pasted from other files.

Reply:
Fixed in d871493